SEGV when attempting to aggregate data JSUP in vector runtime

## tl;dr

#5646 added the ability to read any format in the vector runtime. However, when I just attempted it with JSUP, there was a SEGV.

With test data `data.jsup`:

```
{ResolutionWidth: 0 (int16)}
{ResolutionWidth: 11492 (int16)}
```

Repro:

```
$ SUPER_VAM=1 super -c "from 'data.jsup' | max(ResolutionWidth)"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x7af38cf]

goroutine 14 [running]:
github.com/brimdata/super/zbuf.(*puller).Pull(0xc0004827c0, 0x0?)
	/Users/phil/work/super/zbuf/batch.go:88 +0x6f
github.com/brimdata/super/zbuf.(*namedScanner).Pull(0xc00089cdc0, 0x0?)
	/Users/phil/work/super/zbuf/scanner.go:167 +0x27
github.com/brimdata/super/zbuf.(*namedScanner).Pull(0xc00089cde0, 0x0?)
	/Users/phil/work/super/zbuf/scanner.go:167 +0x27
github.com/brimdata/super/runtime/exec.(*closePuller).Pull(0xc00089ce00, 0x0?)
	/Users/phil/work/super/runtime/exec/environment.go:121 +0x23
github.com/brimdata/super/runtime/vam.(*dematerializer).Pull(0x0?, 0x0?)
	/Users/phil/work/super/runtime/vam/materialize.go:58 +0x42
github.com/brimdata/super/runtime/vam/op/summarize.(*Summarize).Pull(0xc0008b0bb0, 0x0?)
	/Users/phil/work/super/runtime/vam/op/summarize/summarize.go:58 +0x155
github.com/brimdata/super/runtime/vam/op.(*combineParent).run(0xc000896e40)
	/Users/phil/work/super/runtime/vam/op/combine.go:110 +0x3f
created by github.com/brimdata/super/runtime/vam/op.(*Combine).Pull.func1 in goroutine 1
	/Users/phil/work/super/runtime/vam/op/combine.go:42 +0x35
```

## Details

Repro is with super commit 0617b77. I bumped into this while working on variations of the [ClickBench q29 query](https://github.com/ClickHouse/ClickBench/blob/acdd156c11043d45e9e54f791747bf51e9564d06/duckdb/queries.sql#L30).

FWIW, I can read the file and do value-at-a-time operations without complaint.

```
$ super -version
Version: v1.18.0-274-g0617b774

$ SUPER_VAM=1 super -c "from 'data.jsup'"
{ResolutionWidth:0(int16)}
{ResolutionWidth:11492(int16)}

$ SUPER_VAM=1 super -c "from 'data.jsup' | yield ResolutionWidth+1"
1
11493
```

But the problem comes up with all the aggregations I've tried. At the top of the issue I showed a repro with `max()`, but I can also do:

```
$ SUPER_VAM=1 super -c "from 'data.jsup' | sum(ResolutionWidth)"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x5b748cf]

goroutine 59 [running]:
github.com/brimdata/super/zbuf.(*puller).Pull(0xc00075af90, 0x0?)
	/Users/phil/work/super/zbuf/batch.go:88 +0x6f
github.com/brimdata/super/zbuf.(*namedScanner).Pull(0xc00075e620, 0x0?)
	/Users/phil/work/super/zbuf/scanner.go:167 +0x27
github.com/brimdata/super/zbuf.(*namedScanner).Pull(0xc00075e640, 0x0?)
	/Users/phil/work/super/zbuf/scanner.go:167 +0x27
github.com/brimdata/super/runtime/exec.(*closePuller).Pull(0xc00075e660, 0x0?)
	/Users/phil/work/super/runtime/exec/environment.go:121 +0x23
github.com/brimdata/super/runtime/vam.(*dematerializer).Pull(0x0?, 0x0?)
	/Users/phil/work/super/runtime/vam/materialize.go:58 +0x42
github.com/brimdata/super/runtime/vam/op/summarize.(*Sum).Pull(0xc00072d270, 0x0?)
	/Users/phil/work/super/runtime/vam/op/summarize/agg.go:307 +0x73
github.com/brimdata/super/runtime/vam/op.(*combineParent).run(0xc00073d540)
	/Users/phil/work/super/runtime/vam/op/combine.go:110 +0x3f
created by github.com/brimdata/super/runtime/vam/op.(*Combine).Pull.func1 in goroutine 1
	/Users/phil/work/super/runtime/vam/op/combine.go:42 +0x35
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x5b748cf]

goroutine 60 [running]:
github.com/brimdata/super/zbuf.(*puller).Pull(0xc00075af90, 0x0?)
	/Users/phil/work/super/zbuf/batch.go:88 +0x6f
github.com/brimdata/super/zbuf.(*namedScanner).Pull(0xc00075e620, 0x0?)
	/Users/phil/work/super/zbuf/scanner.go:167 +0x27
github.com/brimdata/super/zbuf.(*namedScanner).Pull(0xc00075e640, 0x0?)
	/Users/phil/work/super/zbuf/scanner.go:167 +0x27
github.com/brimdata/super/runtime/exec.(*closePuller).Pull(0xc00075e660, 0x0?)
	/Users/phil/work/super/runtime/exec/environment.go:121 +0x23
github.com/brimdata/super/runtime/vam.(*dematerializer).Pull(0x0?, 0x0?)
	/Users/phil/work/super/runtime/vam/materialize.go:58 +0x42
github.com/brimdata/super/runtime/vam/op/summarize.(*Sum).Pull(0xc00072d2c0, 0x0?)
	/Users/phil/work/super/runtime/vam/op/summarize/agg.go:307 +0x73
github.com/brimdata/super/runtime/vam/op.(*combineParent).run(0xc00073d580)
	/Users/phil/work/super/runtime/vam/op/combine.go:110 +0x3f
created by github.com/brimdata/super/runtime/vam/op.(*Combine).Pull.func1 in goroutine 1
	/Users/phil/work/super/runtime/vam/op/combine.go:42 +0x35
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x5b748cf]

goroutine 58 [running]:
github.com/brimdata/super/zbuf.(*puller).Pull(0xc00075af90, 0x0?)
	/Users/phil/work/super/zbuf/batch.go:88 +0x6f
github.com/brimdata/super/zbuf.(*namedScanner).Pull(0xc00075e620, 0x0?)
	/Users/phil/work/super/zbuf/scanner.go:167 +0x27
github.com/brimdata/super/zbuf.(*namedScanner).Pull(0xc00075e640, 0x0?)
	/Users/phil/work/super/zbuf/scanner.go:167 +0x27
github.com/brimdata/super/runtime/exec.(*closePuller).Pull(0xc00075e660, 0x0?)
	/Users/phil/work/super/runtime/exec/environment.go:121 +0x23
github.com/brimdata/super/runtime/vam.(*dematerializer).Pull(0x0?, 0x0?)
	/Users/phil/work/super/runtime/vam/materialize.go:58 +0x42
github.com/brimdata/super/runtime/vam/op/summarize.(*Sum).Pull(0xc00072d220, 0x0?)
	/Users/phil/work/super/runtime/vam/op/summarize/agg.go:307 +0x73
github.com/brimdata/super/runtime/vam/op.(*combineParent).run(0xc00073d500)
	/Users/phil/work/super/runtime/vam/op/combine.go:110 +0x3f
created by github.com/brimdata/super/runtime/vam/op.(*Combine).Pull.func1 in goroutine 1
	/Users/phil/work/super/runtime/vam/op/combine.go:42 +0x35
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x5b748cf]

goroutine 61 [running]:
github.com/brimdata/super/zbuf.(*puller).Pull(0xc00075af90, 0x0?)
	/Users/phil/work/super/zbuf/batch.go:88 +0x6f
github.com/brimdata/super/zbuf.(*namedScanner).Pull(0xc00075e620, 0x0?)
	/Users/phil/work/super/zbuf/scanner.go:167 +0x27
github.com/brimdata/super/zbuf.(*namedScanner).Pull(0xc00075e640, 0x0?)
	/Users/phil/work/super/zbuf/scanner.go:167 +0x27
github.com/brimdata/super/runtime/exec.(*closePuller).Pull(0xc00075e660, 0x0?)
	/Users/phil/work/super/runtime/exec/environment.go:121 +0x23
github.com/brimdata/super/runtime/vam.(*dematerializer).Pull(0x0?, 0x0?)
	/Users/phil/work/super/runtime/vam/materialize.go:58 +0x42
github.com/brimdata/super/runtime/vam/op/summarize.(*Sum).Pull(0xc00072d310, 0x0?)
	/Users/phil/work/super/runtime/vam/op/summarize/agg.go:307 +0x73
github.com/brimdata/super/runtime/vam/op.(*combineParent).run(0xc00073d5c0)
	/Users/phil/work/super/runtime/vam/op/combine.go:110 +0x3f
created by github.com/brimdata/super/runtime/vam/op.(*Combine).Pull.func1 in goroutine 1
	/Users/phil/work/super/runtime/vam/op/combine.go:42 +0x35
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x5b748cf]

goroutine 57 [running]:
github.com/brimdata/super/zbuf.(*puller).Pull(0xc00075af90, 0x0?)
	/Users/phil/work/super/zbuf/batch.go:88 +0x6f
github.com/brimdata/super/zbuf.(*namedScanner).Pull(0xc00075e620, 0x0?)
	/Users/phil/work/super/zbuf/scanner.go:167 +0x27
github.com/brimdata/super/zbuf.(*namedScanner).Pull(0xc00075e640, 0x0?)
	/Users/phil/work/super/zbuf/scanner.go:167 +0x27
github.com/brimdata/super/runtime/exec.(*closePuller).Pull(0xc00075e660, 0x0?)
	/Users/phil/work/super/runtime/exec/environment.go:121 +0x23
github.com/brimdata/super/runtime/vam.(*dematerializer).Pull(0x0?, 0x0?)
	/Users/phil/work/super/runtime/vam/materialize.go:58 +0x42
github.com/brimdata/super/runtime/vam/op/summarize.(*Sum).Pull(0xc00072d1d0, 0x0?)
	/Users/phil/work/super/runtime/vam/op/summarize/agg.go:307 +0x73
github.com/brimdata/super/runtime/vam/op.(*combineParent).run(0xc00073d4c0)
	/Users/phil/work/super/runtime/vam/op/combine.go:110 +0x3f
created by github.com/brimdata/super/runtime/vam/op.(*Combine).Pull.func1 in goroutine 1
	/Users/phil/work/super/runtime/vam/op/combine.go:42 +0x35
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x5b748cf]

goroutine 53 [running]:
github.com/brimdata/super/zbuf.(*puller).Pull(0xc00075af90, 0x1?)
	/Users/phil/work/super/zbuf/batch.go:88 +0x6f
github.com/brimdata/super/zbuf.(*namedScanner).Pull(0xc00075e620, 0x0?)
	/Users/phil/work/super/zbuf/scanner.go:167 +0x27
github.com/brimdata/super/zbuf.(*namedScanner).Pull(0xc00075e640, 0x0?)
	/Users/phil/work/super/zbuf/scanner.go:167 +0x27
github.com/brimdata/super/runtime/exec.(*closePuller).Pull(0xc00075e660, 0xda?)
	/Users/phil/work/super/runtime/exec/environment.go:121 +0x23
github.com/brimdata/super/runtime/vam.(*dematerializer).Pull(0x1?, 0x0?)
	/Users/phil/work/super/runtime/vam/materialize.go:58 +0x42
github.com/brimdata/super/runtime/vam/op/summarize.(*Sum).Pull(0xc00072d090, 0x1?)
	/Users/phil/work/super/runtime/vam/op/summarize/agg.go:307 +0x73
github.com/brimdata/super/runtime/vam/op.(*combineParent).run(0xc00073d380)
	/Users/phil/work/super/runtime/vam/op/combine.go:110 +0x3f
created by github.com/brimdata/super/runtime/vam/op.(*Combine).Pull.func1 in goroutine 1
	/Users/phil/work/super/runtime/vam/op/combine.go:42 +0x35
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x5b748cf]

goroutine 63 [running]:
github.com/brimdata/super/zbuf.(*puller).Pull(0xc00075af90, 0x0?)
	/Users/phil/work/super/zbuf/batch.go:88 +0x6f
github.com/brimdata/super/zbuf.(*namedScanner).Pull(0xc00075e620, 0x0?)
	/Users/phil/work/super/zbuf/scanner.go:167 +0x27
github.com/brimdata/super/zbuf.(*namedScanner).Pull(0xc00075e640, 0x0?)
	/Users/phil/work/super/zbuf/scanner.go:167 +0x27
github.com/brimdata/super/runtime/exec.(*closePuller).Pull(0xc00075e660, 0x0?)
	/Users/phil/work/super/runtime/exec/environment.go:121 +0x23
github.com/brimdata/super/runtime/vam.(*dematerializer).Pull(0x0?, 0x0?)
	/Users/phil/work/super/runtime/vam/materialize.go:58 +0x42
github.com/brimdata/super/runtime/vam/op/summarize.(*Sum).Pull(0xc00072d3b0, 0x0?)
	/Users/phil/work/super/runtime/vam/op/summarize/agg.go:307 +0x73
github.com/brimdata/super/runtime/vam/op.(*combineParent).run(0xc00073d640)
	/Users/phil/work/super/runtime/vam/op/combine.go:110 +0x3f
created by github.com/brimdata/super/runtime/vam/op.(*Combine).Pull.func1 in goroutine 1
	/Users/phil/work/super/runtime/vam/op/combine.go:42 +0x35
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x5b748cf]

goroutine 55 [running]:
github.com/brimdata/super/zbuf.(*puller).Pull(0xc00075af90, 0x0?)
	/Users/phil/work/super/zbuf/batch.go:88 +0x6f
github.com/brimdata/super/zbuf.(*namedScanner).Pull(0xc00075e620, 0x0?)
	/Users/phil/work/super/zbuf/scanner.go:167 +0x27
github.com/brimdata/super/zbuf.(*namedScanner).Pull(0xc00075e640, 0x0?)
	/Users/phil/work/super/zbuf/scanner.go:167 +0x27
github.com/brimdata/super/runtime/exec.(*closePuller).Pull(0xc00075e660, 0x0?)
	/Users/phil/work/super/runtime/exec/environment.go:121 +0x23
github.com/brimdata/super/runtime/vam.(*dematerializer).Pull(0x0?, 0x0?)
	/Users/phil/work/super/runtime/vam/materialize.go:58 +0x42
github.com/brimdata/super/runtime/vam/op/summarize.(*Sum).Pull(0xc00072d130, 0x0?)
	/Users/phil/work/super/runtime/vam/op/summarize/agg.go:307 +0x73
github.com/brimdata/super/runtime/vam/op.(*combineParent).run(0xc00073d400)
	/Users/phil/work/super/runtime/vam/op/combine.go:110 +0x3f
created by github.com/brimdata/super/runtime/vam/op.(*Combine).Pull.func1 in goroutine 1
	/Users/phil/work/super/runtime/vam/op/combine.go:42 +0x35
```

Or:

```
$ SUPER_VAM=1 super -c "from 'data.jsup' | avg(ResolutionWidth)"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0xe91e8cf]

goroutine 68 [running]:
github.com/brimdata/super/zbuf.(*puller).Pull(0xc0008101f0, 0x0?)
	/Users/phil/work/super/zbuf/batch.go:88 +0x6f
github.com/brimdata/super/zbuf.(*namedScanner).Pull(0xc000812200, 0x0?)
	/Users/phil/work/super/zbuf/scanner.go:167 +0x27
github.com/brimdata/super/zbuf.(*namedScanner).Pull(0xc000812220, 0x0?)
	/Users/phil/work/super/zbuf/scanner.go:167 +0x27
github.com/brimdata/super/runtime/exec.(*closePuller).Pull(0xc000812240, 0x0?)
	/Users/phil/work/super/runtime/exec/environment.go:121 +0x23
github.com/brimdata/super/runtime/vam.(*dematerializer).Pull(0x0?, 0x0?)
	/Users/phil/work/super/runtime/vam/materialize.go:58 +0x42
github.com/brimdata/super/runtime/vam/op/summarize.(*Summarize).Pull(0xc00084e2c0, 0x0?)
	/Users/phil/work/super/runtime/vam/op/summarize/summarize.go:58 +0x155
github.com/brimdata/super/runtime/vam/op.(*combineParent).run(0xc00081c500)
	/Users/phil/work/super/runtime/vam/op/combine.go:110 +0x3f
created by github.com/brimdata/super/runtime/vam/op.(*Combine).Pull.func1 in goroutine 1
	/Users/phil/work/super/runtime/vam/op/combine.go:42 +0x35
```

Now that I'm finishing this write-up, I see this may share a root cause with #5644. If I repeat that last repro several times, I can also see these variations:

```
$ SUPER_VAM=1 super -c "from 'data.jsup' | avg(ResolutionWidth)"
data.jsup: data.jsup: Super JSON syntax error

$ SUPER_VAM=1 super -c "from 'data.jsup' | avg(ResolutionWidth)"
data.jsup: data.jsup: read /Users/phil/work/newbug/data.jsup: file already closed
```

--- a/runtime/vam/materialize.go
+++ b/runtime/vam/materialize.go
@@ -2,6 +2,7 @@ package vam
 
 import (
 	"bytes"
+	"sync"
 
 	"github.com/brimdata/super"
 	"github.com/brimdata/super/vector"
@@ -47,15 +48,18 @@ func (m *Materializer) Pull(done bool) (zbuf.Batch, error) {
 }
 
 type dematerializer struct {
+	mu     sync.Mutex
 	parent zbuf.Puller
 }
 
 func NewDematerializer(p zbuf.Puller) vector.Puller {
-	return &dematerializer{p}
+	return &dematerializer{parent: p}
 }
 
 func (d *dematerializer) Pull(done bool) (vector.Any, error) {
+	d.mu.Lock()
 	batch, err := d.parent.Pull(done)
+	d.mu.Unlock()
 	if batch == nil || err != nil {
 		return nil, err
 	}

